include/etcpal/cpp/inet.h

```diff
@@ -0,0 +1,658 @@
+/******************************************************************************
+ * Copyright 2019 ETC Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************
+ * This file is a part of EtcPal. For more information, go to:
+ * https://github.com/ETCLabs/EtcPal
+ ******************************************************************************/
+
+/// \file etcpal/cpp/inet.h
+/// \brief C++ wrapper and utilities for etcpal/inet.h
+
+#ifndef ETCPAL_CPP_INET_H_
+#define ETCPAL_CPP_INET_H_
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iterator>
+#include <string>
+#include "etcpal/inet.h"
+
+namespace etcpal
+{
+/// \defgroup etcpal_cpp_inet etcpal_cpp_inet
+/// \ingroup etcpal_cpp
+/// \brief C++ utilities for the \ref etcpal_inet module.
+
+/// \ingroup etcpal_cpp_inet
+/// \brief A wrapper class for the EtcPal IP address type.
+///
+/// Provides C++ syntactic sugar for working with IP addresses.
+class IpAddr
+{
+public:
+  IpAddr() noexcept;
+  // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
+  //   etcpal::IpAddr ip = etcpal_c_function_that_returns_ip();
+  IpAddr(const EtcPalIpAddr& c_ip) noexcept;
+  IpAddr& operator=(const EtcPalIpAddr& c_ip) noexcept;
+
+  IpAddr(uint32_t v4_data) noexcept;
+  IpAddr(const uint8_t* v6_data) noexcept;
+  IpAddr(const uint8_t* v6_data, unsigned long scope_id) noexcept;
+
+  const EtcPalIpAddr& get() const noexcept;
+  std::string ToString() const;
+  uint32_t v4_data() const noexcept;
+  const uint8_t* v6_data() const noexcept;
+  unsigned long scope_id() const noexcept;
+
+  bool IsValid() const noexcept;
+  etcpal_iptype_t type() const noexcept;
```

Comment:
Would it be worthwhile to have an enum class for these types to have the nice scoping of the values?

Reply:
Strong typing would be nice

Reply:
So like...
```C++
class IpAddr
{
public:
  enum class Type
  {
    Invalid = kEtcPalIpTypeInvalid,
    V4 = kEtcPalIpTypeV4,
    V6 = kEtcPalIpTypeV6
  };
  ...
  Type type() const noexcept;
  ...
};
...
inline IpAddr::Type type() const noexcept
{
  return static_cast<Type>(addr_.type);
}
...
if (addr.type() == etcpal::IpAddr::Type::V4)
{
  ...
}
```

Something like that? My only issue is it's kind of a lot of `::`s in the usage...

Reply:
Hmm, that does end up being a lot of `::`. I wonder about having an `IpAddrType` class not nested so it's just `etcpal::IpAddrType::V4` instead of the extra layer. Though that's not as nice feeling to me.